src/main/java/com/amazon/ion/impl/macro/MacroCompiler.kt

```diff
@@ -0,0 +1,240 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.amazon.ion.impl.macro
+
+import com.amazon.ion.*
+import com.amazon.ion.impl.macro.TemplateBodyExpression.*
+import com.amazon.ion.util.confirm
+
+/**
+ * [MacroCompiler] wraps an [IonReader]. When directed to do so, it will take over advancing and getting values from the
+ * reader in order to read one [TemplateMacro].
+ *
+ * This is currently implemented using [IonReader], but it could be adapted to work with
+ * [IonReaderContinuableCore][com.amazon.ion.impl.IonReaderContinuableCore].
+ */
+class MacroCompiler(private val reader: IonReader) {
+    // TODO: Make sure that we can throw exceptions if there's an over-sized value.
+
+    /** The name of the macro that was read. Returns `null` if no macro name is available. */
+    var macroName: String? = null
+        private set // Only mutable internally
+
+    private val signature: MutableList<Macro.Parameter> = mutableListOf()
+    private val expressions: MutableList<TemplateBodyExpression> = mutableListOf()
+
+    /**
+     * Compiles a template macro definition from the reader. Caller is responsible for positioning the reader at—but not
+     * stepped into—the macro template s-expression.
+     *
+     * TODO: if we switch the macro compiler to use a continuable reader, change the return type of this
+     *       to a compiler state enum, and add a separate function to get the compiled macro once it is ready.
+     */
+    fun compileMacro(): TemplateMacro {
+        macroName = null
+        signature.clear()
+        expressions.clear()
+
+        confirm(reader.type == IonType.SEXP) { "macro compilation expects a sexp starting with the keyword `macro`" }
+        reader.confirmNoAnnotations("a macro definition sexp")
+        reader.readContainer {
+            confirm(reader.next() == IonType.SYMBOL && reader.stringValue() == "macro") { "macro compilation expects a sexp starting with the keyword `macro`" }
+
+            nextAndCheckType(IonType.SYMBOL, "macro name")
+            confirmNoAnnotations("macro name")
+            // TODO: Enforce 'identifier' syntax subset of symbol
+            //       Possibly add support for macro definitions without names?
+            macroName = symbolValue().assumeText()
+
+            nextAndCheckType(IonType.SEXP, "macro signature")
+            confirmNoAnnotations("macro signature")
+            readSignature()
+
+            forEachRemaining { compileTemplateBodyExpression(isQuoted = false) }
```

Comment:
My understanding had been that the template body must be a single expression. Here's the grammar I'm looking at: https://github.com/amazon-ion/ion-docs/blob/ion-11-specification/src/grammar.adoc#macro-definitions
Am I misreading the grammar, or is something out of date?

Reply:
I'll have to revisit that.

Reply:
I was wrong about this.